keps/sig-cli/2551-return-code-normalization/README.md

```diff
@@ -0,0 +1,652 @@
+only when the pipeline breaks because of a specific error code that represents this authorization failure.
+
+#### Story 3
+Roberta works as the product manager of a big CI/CD SaaS provider. They want to have in their marketplace the
+execution of kubernetes commands targeting a cluster, and fastly providing a feedback to the user if the error
+was due to something on the client side (like a missing flag, an invalid yaml file because of...tabs...) or if this
+is due to some invalid operation on the server side.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+* Users already relying on specific error codes might face migration problems / false positives or negatives 
+because of some specific error code change. The proposal here is to decide if the old or new return code behavior
+will be used based on a well defined environment variable. The old behavior will be kept and followed by the 
+deprecation path will be defaulted after a number of releases.
+
+```
+<<[UNRESOLVED feature gating ]>>
+Decide if a global flag to control this behavior will be present, or if the Env Variable is enough
+<<[/UNRESOLVED]>>
+```
+
+
+* Some exit code might not be proper related to an error. The reserved exit codes are documented [here](https://tldp.org/LDP/abs/html/exitcodes.html)
+and should be used carefully in a way to not generate conflict with existing scripts. 
+
+```
+<<[UNRESOLVED initial error codes ]>>
+Define the initial error codes, and how to document them
+<<[/UNRESOLVED]>>
+```
+
+## Design Details
+
+The majority of commands already are organized as the following:
+* Run Complete to complete missing information with defaults. This runs on the client side
+* Run Validation to check command syntax and missing arguments. This runs on the client side
+* Run the command itself. This might run on the client side, be dry-run, run an external command or call the APIServer.
+
+One thing that should be done is map all the error codes/ints as constants in some file, so they can be automatically
+documented.
+
+There are three design suggestions:
+
+```
+<<[UNRESOLVED select design approach ]>>
+Select the best design approach, move others to design alternatives
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED dealing with kubectl exec and run exit ]>>
+kubectl exec and run uses the pod exit code as its own exit code, we should figure out how we 
+should deal with this
+https://github.com/kubernetes/kubernetes/blob/v1.22.0-alpha.0/test/e2e/kubectl/kubectl.go#L496
+https://github.com/kubernetes/kubernetes/blob/v1.22.0-alpha.0/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L178-L179
```

Comment:
Yeah, that's a big one and how we can properly handle that.

Reply:
Should this follow the 128+exit code rule proposed for plugins?

Reply:
One other option is to start known kubectl error codes at 200. This would allow exec to continue returning the pod exit code, assuming it's less than 200. If the pod exit code is 200 or greater, then set it to 255. Maybe we should apply the same logic to external commands as well, such as diff to keep it consistent. Thoughts on that approach?

Reply:
How about having `kubectl exec` take a parameter `--use-exit-code` that defaults to true, and that you can force to false if you want success to mean “we were able to `exec` _something_”?